packages/world/src/systems/libraries/MoveLib.sol

```diff
@@ -119,73 +154,92 @@ library MoveLib {
       && !EntityUtils.getMovableEntityAt(belowCoord)._exists() && !_isFluid(playerCoord);
   }
 
-  function _computeGravityResult(Vec3 coord, uint16 initialFallHeight) private view returns (Vec3, uint128) {
-    uint16 currentFallHeight = initialFallHeight;
-    Vec3 current = coord;
-
+  function _computeGravityResult(MoveContext memory ctx, Vec3 current, uint16 currentFallHeight)
+    private
+    view
+    returns (Vec3, StepContext memory, uint128)
+  {
     while (_gravityApplies(current)) {
       current = current - vec3(0, 1, 0);
       unchecked {
         ++currentFallHeight;
       }
     }
 
-    // We don't count move units as we are not moving, just falling
-    (uint128 moveCost,) = _getMoveCost(current);
-    // If currently on water or under the safe fall threshold, don't apply fall damage
-    if (currentFallHeight <= Constants.PLAYER_SAFE_FALL_DISTANCE || _isFluid(current)) {
-      return (current, moveCost);
+    // Move step on landing (discounted)
+    StepContext memory stepCtx = _stepContext(ctx, current);
+
+    return (current, stepCtx, _computeFallDamage(ctx, stepCtx, currentFallHeight));
+  }
+
+  function _clampEnergyCosts(MoveContext memory ctx, uint128 currentCost, uint128 moveCost, uint128 fallCost)
+    internal
+    pure
+    returns (uint128, uint128)
+  {
+    if (ctx.initialEnergy <= currentCost) {
```

Comment:
does this need to be `<` not `<=` since if its equal we still want to bring it to 0?

Reply:
ahh I see, makes sense